### ISSUE TITLE
Tflite: issue with dilated convolution in a dynamic shape CNN

### ISSUE THREAD

**Issue by arolet** [tensorflow/tensorflow#50448]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 (for model creation)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: OnePlus 8 (Android 11), Sony Xperia 10 (Android 10)
- TensorFlow installed from (source or binary): from pip
- TensorFlow version (use command below): 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71 2.5.0)
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.3
- GPU model and memory: Geforce RTX 3060

**Describe the current behavior**

I'm trying to use a pre-trained model for image processing in my android app. The input shape is [None, None, None, 3]
and the output shape is inferred from the input shape. My issue happens when the model has a convolution layer with 
dilation (same models without dilation work fine).

The android code I use for inference is has the following part:
    
    interpreter.resizeInput(0, new int[]{1, height, width, 3});
    interpreter.allocateTensors();
    Tensor tensor = interpreter.getOutputTensor(0);
    int[] outShape = tensor.shape();

My issue is that the content of outShape is always [1, 1, 1, outChannels]. I'm also not able to recover the output
of the model with interpreter.runForMultipleInputsOutputs even if I provide arrays with the correct output sizes.

**Describe the expected behavior**

The content of shape should depend on the input (in the example I provide it should be [1, height, width, 16])

**Standalone code to reproduce the issue**

Here is a link to minimal models which show the issue I described:

[dilated_conv.zip](https://github.com/tensorflow/tensorflow/files/6712927/dilated_conv.zip)

I've tried 3 different quantization options but it doesn't change anything.

Here is the python code I used to produce the tflite models. 

    import tensorflow as tf
    import numpy as np
    
    
    def convert(model, fname, quantize_level=0):
        converter = tf.lite.TFLiteConverter.from_keras_model(model)
        if quantize_level > 0:
            converter.optimizations = [tf.lite.Optimize.DEFAULT]
            if quantize_level > 1:
                converter.target_spec.supported_types = [tf.float16]
                fname = "{}_quantized_f16".format(fname)
            else:
                fname = "{}_quantized_default".format(fname)
    
        tflite_model = converter.convert()
        with open("{}.tflite".format(fname), 'wb') as f:
            f.write(tflite_model)
    
    
    x = tf.keras.layers.Input([None, None, 3], dtype=tf.float32)
    conv = tf.keras.layers.Conv2D(filters=16, kernel_size=(3, 3), padding="same", dilation_rate=6, activation=None)
    conv.build(x.shape)
    conv.set_weights([np.random.rand(3, 3, 3, 16), np.random.rand(16)])
    y = conv(x)
    model = tf.keras.Model(x, y)
    
    for i in range(3):
        convert(model, "dilated_conv", quantize_level=i)

I have also tried to add a reshape layer but it doesn't make any difference:

    x = tf.keras.layers.Input([None, None, 3], dtype=tf.float32)
    conv = tf.keras.layers.Conv2D(filters=16, kernel_size=(3, 3), padding="same", dilation_rate=6, activation=None)
    conv.build(x.shape)
    conv.set_weights([np.random.rand(3, 3, 3, 16), np.random.rand(16)])
    y = conv(x)
    y = tf.reshape(y, tf.shape(x))
    model = tf.keras.Model(x, y)

For the android part, I am using the java API and the issue is there with tflite versions 2.2.0, 2.3.0 and 2.5.0.
Just creating an interpreter from one of the above models (without GPU or NNApi delegate) and trying to resize the 
input shows the behavior I described. (if you feed it an image it will do some computations and produce an output with the
wrong shape)

**Comment by miaout17** [comment]
@haozha111 Do you have time to take a look? or feel free to rebalance this to other people. Thanks

**Comment by haozha111** [comment]
I think this might be an issue happened during shape propagation after you call `ResizeInput`. Have you tried to set the input shape as a static shape (instead of `None` in the dimensions) and see whether the output tensor has the expected shape?

**Comment by arolet** [comment]
Yes, my issue is specifically for dynamic shapes. If I use static shapes to prepare the model and then convert it it works*, but I need dynamic shapes in my app.

* actually if I set a static shape with the batch size set to 1, I get a crash on the default delegate and the NNApi delegate, but it works well on the GPU one. Batch size set to None works with all delegate but is then slower on the GPU one. I feel like this is a different issue though and it is less critical to my use case. Should I report this as a separate issue?

**Comment by arolet** [comment]
@haozha111 Any idea why this happens? By the way the problem seems to be specific to the android tflite library, and not to the model itself. I tried to run the same model with the python interpreter and it works as expected.

**Comment by haozha111** [comment]
Thanks for the info. If the issue only happens in the android tflite library, and not reproducible on desktop with the python interpreter, then I guess it's probably not caused by the dilated conv in the model. Have you tried the latest tf-nightly and see if the problem fixes?

**Comment by arolet** [comment]
Thank you for the answer.

I just tried with tf-nightly and it gives the same results.

> it's probably not caused by the dilated conv in the model

I'm not sure what you mean by that, but the issue doesn't happen with other layers. In the minimal model I provided, which is only one convolutional layer, if you remove dilation then the bug doesn't happen. Similarly in my actual model (which has many more layers), if the bug only happens if there is some dilation somewhere.

What's weird is that I have tried adding a reshape layer after the dilated convolution to enforce the correct shape but it doesn't change anything.

**Comment by haozha111** [comment]
Sure. So the issue is gone if you use the python interpreter (on desktop) ?

**Comment by arolet** [comment]
Yes, the python interpreter has the correct shape, and give the same output as the original tensorflow (not lite) model.

**Comment by haozha111** [comment]
Hi,

I checked the float tflite model you shared, and I found that the dilated conv pattern is not folded (there is still the SpaceToBatchNd, and BatchToSpaceNd op in your model). This is because the BatchToSpaceNd op is taking > 1 input. Given it's not folded, I think the shape issue is not caused by the dilated conv folding pass in MLIR.

I also checked that your model's input has static shape [1, 1, 1, 3] while its shape signature is [-1, -1, -1, 3], this is correct since the first 3 dimensions are dynamic. Given that the output shape is correct when you run in python interpreter, but becomes incorrect when you run on android, I suspect this is caused by the JNI binding layer. When you call Resize in Java, I suspect something is going wrong in the JNI layer which fails to propagate the new shape to each tensor. Could you also share your Java code that resizes the model's input shape?

**Comment by arolet** [comment]
Hi, thank you for the answer.

I tried to extract only the relevant parts of the code, if you need more details please let me know. The models I provided are in the assets folder of the app. Here is the class responsible for running the model on a bitmap.

```
public class TfLiteIssue {

    private final static String TAG = "TfLiteIssue";
    private final Interpreter interpreter;

    public TfLiteIssue(Context context) throws IOException {
        this(context, "dilated_conv.tflite");
    }

    public TfLiteIssue(Context context, String modelPath) throws IOException {
        AssetFileDescriptor fileDescriptor = context.getAssets().openFd(modelPath);
        FileInputStream inputStream = new FileInputStream(fileDescriptor.getFileDescriptor());
        long startOffset = fileDescriptor.getStartOffset();
        long declaredLength = fileDescriptor.getDeclaredLength();
        FileChannel fileChannel = inputStream.getChannel();
        interpreter = new Interpreter(fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength),
                new Interpreter.Options());
    }


    public float[][][][] runModel(Bitmap input) {
        interpreter.resizeInput(0, new int[]{1, input.getHeight(), input.getWidth(), 3});
        interpreter.allocateTensors();
        Log.d(TAG, "Output shape " + Arrays.toString(interpreter.getOutputTensor(0).shape()));
        Map<Integer, Object> tfOutput = new HashMap<>();
        tfOutput.put(0, createOutputArray());
        interpreter.runForMultipleInputsOutputs(new float[][][][][]{asFloatArray(input)}, tfOutput);
        return (float[][][][]) tfOutput.get(0);
    }


    private float[][][][] createOutputArray() {
        Tensor tensor = interpreter.getOutputTensor(0);
        int[] shape = tensor.shape();
        return new float[1][shape[1]][shape[2]][shape[3]];
    }

    private float[][][][] asFloatArray(Bitmap bitmap) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        int[] intPixels = new int[height * width];
        float[][][][] floatPixels = new float[1][height][width][3];
        bitmap.getPixels(intPixels, 0, width, 0, 0, width, height);
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                int pixel = intPixels[j + i * width];
                floatPixels[0][i][j][0] = (float) (pixel & 0x00FF0000 >> 16) / 255;
                floatPixels[0][i][j][1] = (float) (pixel & 0x0000FF00 >> 8) / 255;
                floatPixels[0][i][j][2] = (float) (pixel & 0x000000FF) / 255;
            }
        }
        return floatPixels;
    }
}
```

In my main activity, I simply call

`TfLiteIssue(getContext()).runModel(bitmap);`

In the logcat this shows an output shape of [1, 1, 1, 3], instead of something with the same shape as the input image.
My bitmap comes from a content provider but it gives the same result with a constant color, for example

```
Bitmap image = Bitmap.createBitmap(256, 256, Bitmap.Config.ARGB_8888);
image.eraseColor(android.graphics.Color.GREEN);
TfLiteIssue(getContext()).runModel(bitmap);
```

**Comment by haozha111** [comment]
@lu-wang-g Hi Lu, since this is related with the Java API, do you know if this is the recommended way to do resizing in Java? Is there any API that the TF Lite support library can provide here? Thanks!

**Comment by lu-wang-g** [comment]
The code snippet from @arolet looks good to me. Through I have no clue yet why this happened. I can reproduce the issue locally. I'll take a closer look at it tomorrow.

**Comment by lu-wang-g** [comment]
Found that the output tensor shape won't be updated until `runForMultipleInputsOutputs` is called. So if you move your log after inference, it will show the correct output shape. I also verified and confirmed this behaviour with the Python API.

Before this issue, I also thought `allocateTensors()` is good enough to update the output tensor shape. And I'm pretty sure it works for some dynamic shape models I've been before. Not sure if certain ops may break the rule. @haozha111 do you have any pointers?

**Comment by haozha111** [comment]
From the code description here:
https://github.com/tensorflow/tensorflow/blob/411666f39a4ed530a9e9f95655de022cb675c655/tensorflow/lite/java/src/main/java/org/tensorflow/lite/InterpreterApi.java#L299

Specifically:
Note that, for graphs with output shapes that are dependent on input *values*, the output shape may not be fully determined until
 running inference.

**Comment by haozha111** [comment]
I think the reason is that, in your model the `SpaceToBatchND` op's padding input is a tensor (int32[2, 2]). The value of the padding tensor will determine the output shape of the `SpaceToBatchND` op, see documentation here on the role of `padding`:
https://www.tensorflow.org/api_docs/python/tf/space_to_batch

So I think your model's actual output shape can only be determined after the first inference run. Could you just get the output shape after inference finishes?

**Comment by arolet** [comment]
@lu-wang-g Thank you for your answer. On my device (a OnePlus 8 with android 11), the shape is still wrong after runForMultipleInputsOutputs. allocateTensors reshapes the outputs correctly with all the layers I tried, only convolution with dilation yields a wrong shape.

@haozha111 It could be a solution, however: 1) it doesn't work on the devices I tried and 2) it would require to run the inference twice, since with the android api we need to give buffers of the correct shape for the output (so the one inference to propagate the shape and one to actually get the output), which is quite slow. Also it should be possible to compute the output shape beforehand, since it's done correctly when there is no dilation and in any case, a convolution with padding='same' is not supposed to change the size.

Also, adding a reshape layer to force the correct shape after the dilated convolution doesn't change the issue (see the initial post). To me it shows that there is probably something more going on doesn't it?

**Comment by haozha111** [comment]
Hi,

When you say that the output shape is still wrong after `runForMultipleInputsOutputs`, what's the output shape you get? Is it different from what the python API is giving to you (supposing python API is working normally).

I'm just curious why the behavior is different between the Java and python API.

**Comment by arolet** [comment]
@haozha111 the "wrong" output shape on android is [1, 1, 1, 16], just like before. In python the output shape is the correct one (i.e. [1, im_height, im_width, 16].

**Comment by lu-wang-g** [comment]
Hi @arolet, here is the code I used to run your model:

```
    Interpreter interpreter = new Interpreter(dilated, new Interpreter.Options());
    int[] inputDims = {1, 10, 10, 3};
    interpreter.resizeInput(0, inputDims);

    interpreter.allocateTensors();
    Map<Integer, Object> tfOutput = new HashMap<>();
    tfOutput.put(0, new float[1][10][10][16]);
    float[][][][] input = new float[1][10][10][3];
    interpreter.runForMultipleInputsOutputs(new float[][][][][]{input}, tfOutput);
    Log.d(TAG, "output shape " + Arrays.toString(interpreter.getOutputTensor(0).shape()));
```

The printed shape is [1, 10, 10, 16], And I verified it using a Pixel 4 device. Please check if the code snippet works for you.

**Comment by arolet** [comment]
Hello @lu-wang-g, thank you for your update.

That's something I actually tried before and I get an IllegalArgumentException with the following message:

`Cannot copy from a TensorFlowLite tensor (Identity) with shape [1, 1, 1, 16] to a Java object with shape [1, 384, 704, 16].`

Just to be sure I tried again with the 2.5.0 and nightly versions and both gave me the error. I only tried with my OnePlus 8 yet, I will check other devices later.

**Comment by lu-wang-g** [comment]
@arolet Do you mind to try the exactly same code I pasted above, and see if it works for you? After that, we can continue debugging your code.

**Comment by arolet** [comment]
@lu-wang-g your exact code works! I have tried to check what the difference with mine is and it looks like the problem only arises with non-square images.

More specifically I created this function:

```
    public float[][][][] runModel(int width, int height){
        Interpreter interpreter = new Interpreter(dilated, new Interpreter.Options());
        int[] inputDims = {1, height, width, 3};
        interpreter.resizeInput(0, inputDims);

        interpreter.allocateTensors();
        Log.d(TAG, "Output shape " + Arrays.toString(interpreter.getOutputTensor(0).shape()));
        Map<Integer, Object> tfOutput = new HashMap<>();
        tfOutput.put(0, new float[1][height][width][16]);
        float[][][][] input = new float[1][height][width][3];
        interpreter.runForMultipleInputsOutputs(new float[][][][][]{input}, tfOutput);
        Log.d(TAG, "output shape " + Arrays.toString(interpreter.getOutputTensor(0).shape()));
        return (float[][][][]) tfOutput.get(0);
    }
```

It gives the crash I described the other day if "width!=height" (width=100 and height=200 in my test). If I replace the first line with dimensions in the wrong order, then there is no crash anymore:
```
    public float[][][][] runModel(int width, int height){
        Interpreter interpreter = new Interpreter(dilated, new Interpreter.Options());
        int[] inputDims = {1, width, height, 3};
        interpreter.resizeInput(0, inputDims);

        interpreter.allocateTensors();
        Log.d(TAG, "Output shape " + Arrays.toString(interpreter.getOutputTensor(0).shape()));
        Map<Integer, Object> tfOutput = new HashMap<>();
        tfOutput.put(0, new float[1][height][width][16]);
        float[][][][] input = new float[1][height][width][3];
        interpreter.runForMultipleInputsOutputs(new float[][][][][]{input}, tfOutput);
        Log.d(TAG, "output shape " + Arrays.toString(interpreter.getOutputTensor(0).shape()));
        return (float[][][][]) tfOutput.get(0);
    }
```
I don't know if this helps. (note that dimensions are inversed only in the first line)

**Comment by lu-wang-g** [comment]
If you remove the `Log` right after `allocateTensors()`, it will work. Calling `interpreter.getOutputTensor(0)` will create the Java output tensor from the native layer immediately. Right after `allocateTensors()`, the output tensor shape remains [1, 1, 1, 16]. It will not be refreshed in `runForMultipleInputsOutputs`. Therefore you ended up with the issue that "Cannot copy from a TensorFlowLite tensor (Identity) with shape [1, 1, 1, 16] to a Java object with shape [1, 100, 200, 16].". If you don't call `interpreter.getOutputTensor(0)` before `runForMultipleInputsOutputs `, `runForMultipleInputsOutputs` will create the Java output tensor from native layer during the process, and by that time, the output tensor shape has been updated to [1,  100, 200, 16].

When you reversed `height` and `width`, you changed the input tensor shape,  meaning the input shape of `runForMultipleInputsOutputs` does not match the input shape set from `resizeInput`. Therefore, `runForMultipleInputsOutputs` triggered an implicit resize and updated both input and output tensor shape.

This is a flaw in the workflow of the Java API, and it's not easy to fix right away. So the short answer is don't trigger `interpreter.getOutputTensor()` before `runForMultipleInputsOutputs`.

**Comment by arolet** [comment]
@lu-wang-g oh it does work without the log, that's great thanks!!!

So as long as I know the output size before-hand it should be fine I guess. I will try with my actual model (for which the output size is not actually the same as the input size) and close the issue if it works.

**Comment by arolet** [comment]
I tried with my actual model, and TL;DR: it now works.

If there is an `getOutputTensor()` anywhere before `runForMultipleInputsOutputs()` though, then I get the error I described earlier, even if it is also before `resizeInput()` and `allocateTensors()`. So basically I had to remove all calls to `getOutputTensor()` (I had some just after the interpreter instantiation to get the number of output channels dynamically).

Thanks a lot for your help, I'm closing the issue since there is a workaround (although I still hope it gets fixed some day).

**Comment by jonpsy** [comment]
@arolet Hi, I'm currently writing a [test case](https://github.com/tensorflow/tflite-support/pull/612#issuecomment-891764933) that checks the ability of the model to re-dim the graph as per the given input. The model you've provided could serve as the testing model.  Do you mind if I use your model?

**Comment by arolet** [comment]
@jonpsy sure go ahead, feel free to use any of the code too.
